main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala

```diff
@@ -60,20 +48,28 @@ object CodeHinter {
     * Computes code quality hints for the given expression `exp0`.
     */
   private def visitExp(exp0: Expression)(implicit root: Root): List[CodeHint] = exp0 match {
-    case Expression.Wild(_, _) => Nil
+    case Expression.Wild(tpe, _) =>
+      visitType(tpe)
 
-    case Expression.Var(_, _, _) => Nil
+    case Expression.Var(_, tpe, _) =>
```

Comment:
This is all the right idea. But I don't think we should be traversing the types in all these expressions, because there's no explicit/syntactic type at a location. For example, if I have

```scala
@Deprecated
pub enum E

def f(e: E): E = e
```

I would expect us to highlight the two use-sites of `E`, but _not_ the use of the variable `e`, since that's just a consequence of the type ascription above.

So in expressions I suggest we only track matches, tags, ascriptions, casts, and lambdas, and any other places where a type actually appears in the source code.

Thoughts?

Reply:
@mlutze is right. I think we should just reuse the same 4-5 lines as in `findEnumReferences`  that you already implemented.

Reply:
Thanks for the review :)

@mlutze I have a doubt, what's an ascription precisely?
I completely agree with you about not highlighting the names of var, indeed I am highlighting the type of the var (I am not passing the location of the variable itself), as you can see from the screenshot :)
I need to add casts, I forgot, thanks for noticing! Probably I am also missing some other I will double check

@magnus-madsen I believe I am not fully following you, I also thought to the index to compute the useSites of an enum and then compute the CodeHints for them simply by writing somthing like:
```scala
private def visitEnum(enum: TypedAst.Enum)(implicit root: Root, index: Index): List[CodeHint] = {
    val useSites = index.usesOf(enum.sym)
    val isDeprecated = enum.ann.exists(ann => ann.name.isInstanceOf[Ast.Annotation.Deprecated])
    val isExperimental = enum.ann.exists(ann => ann.name.isInstanceOf[Ast.Annotation.Experimental])
    useSites.flatMap(use => {
      val deprecated = if (isDeprecated) CodeHint.Deprecated(use) :: Nil else Nil
      val experimental = if (isExperimental) CodeHint.Experimental(use) :: Nil else Nil
      deprecated ++ experimental
    }).toList
  }
```
and calling it at top level, however this approach doesn't seem to work, with my surprise. Is this what you meant? Also I find difficult to debug it, since even when I try to print out info (such the number of useSites) the output is not displayed in my extension host.

Reply:
I think maybe we should use both:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/api/lsp/Index.scala#L109
https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/api/lsp/Index.scala#L114

Reply:
> @mlutze I have a doubt, what's an ascription precisely? I completely agree with you about not highlighting the names of var, indeed I am highlighting the type of the var (I am not passing the location of the variable itself), as you can see from the screenshot :) I need to add casts, I forgot, thanks for noticing! Probably I am also missing some other I will double check

When you highlight the type of the var, sometimes that will have the same location as the var, or some undefined location, because it doesn't correspond to a type we can actually see. I'm afraid we might get highlighting in weird places if we do that, so instead I am proposing only handling explicit types.


```scala
def f(x: Int32): Int32 =
    x: Int32      // this is an ascription (saying "I want x to be treated as an Int32, please check that that is possible")

def g(x: Int32): Int32
    x as Int32    // this is a cast (saying "I want x to be treated as an Int32 and I don't care if the compiler says it's not possible")
```

Reply:
> > @mlutze I have a doubt, what's an ascription precisely? I completely agree with you about not highlighting the names of var, indeed I am highlighting the type of the var (I am not passing the location of the variable itself), as you can see from the screenshot :) I need to add casts, I forgot, thanks for noticing! Probably I am also missing some other I will double check
> 
> When you highlight the type of the var, sometimes that will have the same location as the var, or some undefined location, because it doesn't correspond to a type we can actually see. I'm afraid we might get highlighting in weird places if we do that, so instead I am proposing only handling explicit types.
> 
> ```scala
> def f(x: Int32): Int32 =
>     x: Int32      // this is an ascription (saying "I want x to be treated as an Int32, please check that that is possible")
> 
> def g(x: Int32): Int32
>     x as Int32    // this is a cast (saying "I want x to be treated as an Int32 and I don't care if the compiler says it's not possible")
> ```

Thanks for the explanation, everything it's more clear now :)
So if I understand correctly, when we write something like:
```
let x: Int32 = 0
```
that's a type ascription. this makes everything more clear :)

Then I agree with your point, I will remove the `Var` case.

Reply:
Another question: what's precisely a `Wild` type?

Reply:
A wild type is a type variable that is free/ignored in some way. For example, I could do this:

```scala
myList: List[_] // the _ is wild
```

Which is a type ascription that says `myList` is some kind of list, but doesn't specify what kind.

Reply:
I see thanks! We can add this also to the guide maybe? I couldn't find it

Reply:
> @magnus-madsen That's totally right, but I feel there is still something not working, since for the code posted by @mlutze with the approach above nothing is displayed even if the enum symbol is involved and no tags are. Also another strange fact is that even when I compute the codehints for the return type of a function, the highlight is shown only if the function is actully called

If you create an enum and you right click on a tag and choose "find references" what do you see? I think those are the things that should show up with deprecating warnings.

Reply:
@magnus-madsen  I completely agree with you, however there is something not working properly with the index itself which returns an empty set, you can take a look at the code and screenshot here:

https://github.com/flix/flix/pull/3384

I have noticed a bug also in the way the hints are displayed in the `problems` panel, once the problem is fixed it is doesn't go away and is listed in there. However the highlighting goes away, so I believe it is a problem of the extension, but needs more investigation.